### PR TITLE
feat(ui): TE-1388 enabling drag select to assign daterange to anomalies in create anomaly page

### PR DESCRIPTION
**Pull request by alcatraz627**
### Issue(s)

[TE-1388](https://cortexdata.atlassian.net/browse/TE-1388)

### Description

- Enable the user to drag on the alert chart to select a region of anomalous behaviour as a more intuitive way of manually reporting anomaly
- Anomaly date range selection UI improvements 
- Anomaly date range values are read from the URL initially, and updated to the URL on every change
- Anomaly date range selection always rounds to the next closest start and end timestamp, for both dragging and selecting via date range picker
- Made dimension (enumeration item) selection mandatory to create an anomaly
- Redirecting from an alert's specific dimension will load that specific dimension
- Fixed issue where alert dropdown was going from uncontrolled to controlled input
- Enumeration item is updated in the page URL on every change

### Screenshots

- Report Anomaly - Alert Selected
Screenshot 2023-03-17 at 16 53 53

- Report Anomaly - Need to select dimension
Screenshot 2023-03-17 at 16 55 01

- Alert Page with dimensions
Screenshot 2023-03-17 at 16 54 33

- Alert Chart - Anomaly Date Range Select Mode
Screenshot 2023-03-17 at 16 58 01

- Alert Chart - Zoom Mode
Screenshot 2023-03-17 at 16 58 44

### Video Demo

[Click here to view the Video Demo](https://drive.google.com/file/d/1UB8bEvsTUWUHkgL-Kgx3FJC227fpzh6s/preview)

[TE-1388]: https://cortexdata.atlassian.net/browse/TE-1388?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: thirdeye-ui/src/app/components/alert-view/enumeration-items-table/enumeration-item-row/enumeration-item-row.component.tsx [hunk 1]
```diff
@@ -138,12 +197,28 @@ export const EnumerationItemRow: FunctionComponent<EnumerationItemRowProps> = ({
...
+                                        </>
+                                    )}
+                                    {isExpanded && (
+                                        <>
+                                            <span>
+                                                {t("label.hide-details")}
+                                            </span>
+                                            <ExpandLessIcon />
+                                        </>
+                                    )}
```
Comment (spham92) [inline thirdeye-ui/src/app/components/alert-view/enumeration-items-table/enumeration-item-row/enumeration-item-row.component.tsx:220]: Can be simplified to:

```
<span>
    {t("label.view-details")}
</span>
{!isExpanded && (
    <ExpandMoreIcon />
)}
{isExpanded && (
    <ExpandLessIcon />
)}
```

Comment (alcatraz627) [reply thirdeye-ui/src/app/components/alert-view/enumeration-items-table/enumeration-item-row/enumeration-item-row.component.tsx:220]: The labels are different, `view-details` and `hide-details`, so need to have them in the condition

Comment (spham92) [reply thirdeye-ui/src/app/components/alert-view/enumeration-items-table/enumeration-item-row/enumeration-item-row.component.tsx:220]: Ohhh I see

**Review by spham92 — DISMISSED** [review]

File: thirdeye-ui/src/app/components/anomalies-create/create-anomaly-wizard/create-anomaly-wizard.component.tsx [hunk 2]
```diff
@@ -354,28 +354,114 @@ export const CreateAnomalyWizard: FunctionComponent<CreateAnomalyWizardProps> =
...
-                (v) => v >= thresholdValue
-            );
+            return {
+                start: timestamp[0],
+                end: timestamp[timestamp.length - 1],
+            };
+        }, [evaluation]);
+
+        /** @function Ensure that the date range values are always valid and make sense
+         *  @return *boolean* indicating if the attempted correction was successful or not
```
Comment (spham92) [inline thirdeye-ui/src/app/components/anomalies-create/create-anomaly-wizard/create-anomaly-wizard.component.tsx:379]: I follow `JSDOC` https://jsdoc.app/tags-returns.html

Change to
```
@return {boolean} indicating if the attempted correction was successful or not
```

Comment (alcatraz627) [reply thirdeye-ui/src/app/components/anomalies-create/create-anomaly-wizard/create-anomaly-wizard.component.tsx]: Updated according to the JSDOC standard

File: thirdeye-ui/src/app/components/anomalies-create/create-anomaly-wizard/create-anomaly-wizard.component.tsx [hunk 3]
```diff
@@ -354,28 +354,114 @@ export const CreateAnomalyWizard: FunctionComponent<CreateAnomalyWizardProps> =
...
+            end = timestamp.find((t) => t >= endTimeArg) || end;
+
+            // The anomaly must have a non-zero date range
+            if (start === end) {
+                // If the start AND end lie on the last point,
+                // then move the end ahead by the alert granularity
+                if (end === chartDataBounds.end) {
+                    const alertGranularityString =
+                        alertInsight?.templateWithProperties.metadata
+                            .granularity;
```
Comment (spham92) [inline thirdeye-ui/src/app/components/anomalies-create/create-anomaly-wizard/create-anomaly-wizard.component.tsx:447]: If alertInsight is missing, you can use https://github.com/startreedata/thirdeye/blob/3d194460e662cc7ac0d88d1f9a9d381efe2693cb/thirdeye-ui/src/app/utils/visualization/visualization.util.ts#L149

Comment (alcatraz627) [reply thirdeye-ui/src/app/components/anomalies-create/create-anomaly-wizard/create-anomaly-wizard.component.tsx:450]: I wasn't aware of this. Thanks for the point-out, updated.

**Review by spham92 — APPROVED** [review]

**Pull request merged**